pkg/kubelet/lifecycle/handlers_test.go

```diff
@@ -165,6 +189,492 @@ func TestRunHandlerHttp(t *testing.T) {
+	pod.Spec.Containers = []v1.Container{container}
+	_, err := handlerRunner.Run(containerID, &pod, &container, container.Lifecycle.PostStart)
+
+	if err != nil {
+		t.Errorf("unexpected error: %v", err)
+	}
+	if fakeHTTPDoer.url != "http://foo:8080/bar" {
+		t.Errorf("unexpected url: %s", fakeHTTPDoer.url)
+	}
+	if fakeHTTPDoer.headers["Foo"][0] != "bar" {
+		t.Errorf("missing http header: %s", fakeHTTPDoer.headers)
+	}
+}
+
+func TestRunHandlerHttps(t *testing.T) {
+	fakeHTTPDoer := fakeHTTP{}
+	fakePodStatusProvider := podStatusProviderFunc(func(uid types.UID, name, namespace string) (*kubecontainer.PodStatus, error) {
+		return &kubecontainer.PodStatus{
+			ID:        uid,
+			Name:      name,
+			Namespace: namespace,
+			IPs: []string{
+				"127.0.0.1",
+			},
+		}, nil
+	})
+	handlerRunner := NewHandlerRunner(&fakeHTTPDoer, &fakeContainerCommandRunner{}, fakePodStatusProvider)
+
+	containerID := kubecontainer.ContainerID{Type: "test", ID: "abc1234"}
+	containerName := "containerFoo"
+
+	container := v1.Container{
+		Name: containerName,
+		Lifecycle: &v1.Lifecycle{
+			PostStart: &v1.LifecycleHandler{
+				HTTPGet: &v1.HTTPGetAction{
+					Scheme: v1.URISchemeHTTPS,
+					Host:   "foo",
+					Path:   "bar",
+				},
+			},
+		},
+	}
+	pod := v1.Pod{}
+	pod.ObjectMeta.Name = "podFoo"
+	pod.ObjectMeta.Namespace = "nsFoo"
+	pod.Spec.Containers = []v1.Container{container}
+
+	t.Run("consistent", func(t *testing.T) {
+		container.Lifecycle.PostStart.HTTPGet.Port = intstr.FromString("70")
+		pod.Spec.Containers = []v1.Container{container}
+		_, err := handlerRunner.Run(containerID, &pod, &container, container.Lifecycle.PostStart)
+
+		if err != nil {
+			t.Errorf("unexpected error: %v", err)
+		}
+		if fakeHTTPDoer.url != "https://foo:70/bar" {
+			t.Errorf("unexpected url: %s", fakeHTTPDoer.url)
+		}
+	})
+
+	t.Run("inconsistent", func(t *testing.T) {
+		defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.ConsistentHTTPGetHandlers, false)()
+		container.Lifecycle.PostStart.HTTPGet.Port = intstr.FromString("70")
+		pod.Spec.Containers = []v1.Container{container}
+		_, err := handlerRunner.Run(containerID, &pod, &container, container.Lifecycle.PostStart)
+
+		if err != nil {
+			t.Errorf("unexpected error: %v", err)
+		}
+		if fakeHTTPDoer.url != "http://foo:70/bar" {
+			t.Errorf("unexpected url: %q", fakeHTTPDoer.url)
+		}
+	})
+}
+
+func TestRunHandlerHTTPPort(t *testing.T) {
+	tests := []struct {
+		Name               string
+		FeatureGateEnabled bool
+		Port               intstr.IntOrString
+		ExpectError        bool
+		Expected           string
+	}{
+		{
+			Name:               "consistent/with port",
+			FeatureGateEnabled: true,
+			Port:               intstr.FromString("70"),
+			Expected:           "https://foo:70/bar",
+		}, {
+			Name:               "consistent/without port",
+			FeatureGateEnabled: true,
+			Port:               intstr.FromString(""),
+			ExpectError:        true,
```

Comment:
This doesn't represent a change to probes, it only represents a change to handlers when the feature gate is enabled. 

Given that the goal of this PR is to make probers and handlers consistent and that probers currently require a port, isn't this the desired outcome?

Reply:
I meant that I think this is likely to break existing handlers. Yes it's consistent, but we should be trying to minimize breakage, and i think this is a case where doing something inconsistent to preserve backwards compatibility would be justified. I also think it would be reasonable to change the prober behavior to match the handlers (default to 80 or 443), which would preserve consistency and be unlikely to break anything.

Reply:
Trying to keep probers unchanged while also not requiring a port for handlers would be difficult. I'm ok changing the prober in this case. I've never understood why probes require the port to be set, so my only concern is that there's some subtlety that I'm not aware of.

I'll push a commit this weekend that will not require the port when the feature flag is set. Does that sound like a reasonable way to approach this?

Reply:
I pushed that change.

Reply:
> I also think it would be reasonable to change the prober behavior to match the handlers (default to 80 or 443), which would preserve consistency and be unlikely to break anything.

API validation actually requires a non-zero integer port or non-empty string port name to be specified for both probers and lifecycle hooks, so there's no breakage of API-driven uses of lifecycle hooks if we make lifecycle hooks consistent with probers and require the port (which is my preference).

Verified with this manifest
```yaml
kind: Pod
apiVersion: v1
metadata:
  name: test
spec:
  containers:
  - name: missing-port
    image: busybox
    livenessProbe:
      httpGet:
        path: /
    readinessProbe:
      httpGet:
        path: /
    startupProbe:
      httpGet:
        path: /
    lifecycle:
      preStop:
        httpGet:
          path: /
      postStart:
        httpGet:
          path: /
  - name: zero-port
    image: busybox
    livenessProbe:
      httpGet:
        path: /
        port: 0
    readinessProbe:
      httpGet:
        path: /
        port: 0
    startupProbe:
      httpGet:
        path: /
        port: 0
    lifecycle:
      preStop:
        httpGet:
          path: /
          port: 0
      postStart:
        httpGet:
          path: /
          port: 0
  - name: empty-string-port
    image: busybox
    livenessProbe:
      httpGet:
        path: /
        port: ""
    readinessProbe:
      httpGet:
        path: /
        port: ""
    startupProbe:
      httpGet:
        path: /
        port: ""
    lifecycle:
      preStop:
        httpGet:
          path: /
          port: ""
      postStart:
        httpGet:
          path: /
          port: ""
  - name: int-port
    image: busybox
    livenessProbe:
      httpGet:
        path: /
        port: 1
    readinessProbe:
      httpGet:
        path: /
        port: 1
    startupProbe:
      httpGet:
        path: /
        port: 1
    lifecycle:
      preStop:
        httpGet:
          path: /
          port: 1
      postStart:
        httpGet:
          path: /
          port: 1
  - name: string-port
    image: busybox
    livenessProbe:
      httpGet:
        path: /
        port: "http"
    readinessProbe:
      httpGet:
        path: /
        port: "http"
    startupProbe:
      httpGet:
        path: /
        port: "http"
    lifecycle:
      preStop:
        httpGet:
          path: /
          port: "http"
      postStart:
        httpGet:
          path: /
          port: "http"
```

errors:
```shell
The Pod "test" is invalid:
* spec.containers[0].lifecycle.postStart.httpGet.port: Invalid value: 0: must be between 1 and 65535, inclusive
* spec.containers[0].lifecycle.preStop.httpGet.port: Invalid value: 0: must be between 1 and 65535, inclusive
* spec.containers[0].livenessProbe.httpGet.port: Invalid value: 0: must be between 1 and 65535, inclusive
* spec.containers[0].readinessProbe.httpGet.port: Invalid value: 0: must be between 1 and 65535, inclusive
* spec.containers[0].startupProbe.httpGet.port: Invalid value: 0: must be between 1 and 65535, inclusive
* spec.containers[1].lifecycle.postStart.httpGet.port: Invalid value: 0: must be between 1 and 65535, inclusive
* spec.containers[1].lifecycle.preStop.httpGet.port: Invalid value: 0: must be between 1 and 65535, inclusive
* spec.containers[1].livenessProbe.httpGet.port: Invalid value: 0: must be between 1 and 65535, inclusive
* spec.containers[1].readinessProbe.httpGet.port: Invalid value: 0: must be between 1 and 65535, inclusive
* spec.containers[1].startupProbe.httpGet.port: Invalid value: 0: must be between 1 and 65535, inclusive
* spec.containers[2].lifecycle.postStart.httpGet.port: Invalid value: "": must contain only alpha-numeric characters (a-z, 0-9), and hyphens (-)
* spec.containers[2].lifecycle.postStart.httpGet.port: Invalid value: "": must contain at least one letter (a-z)
* spec.containers[2].lifecycle.preStop.httpGet.port: Invalid value: "": must contain only alpha-numeric characters (a-z, 0-9), and hyphens (-)
* spec.containers[2].lifecycle.preStop.httpGet.port: Invalid value: "": must contain at least one letter (a-z)
* spec.containers[2].livenessProbe.httpGet.port: Invalid value: "": must contain only alpha-numeric characters (a-z, 0-9), and hyphens (-)
* spec.containers[2].livenessProbe.httpGet.port: Invalid value: "": must contain at least one letter (a-z)
* spec.containers[2].readinessProbe.httpGet.port: Invalid value: "": must contain only alpha-numeric characters (a-z, 0-9), and hyphens (-)
* spec.containers[2].readinessProbe.httpGet.port: Invalid value: "": must contain at least one letter (a-z)
* spec.containers[2].startupProbe.httpGet.port: Invalid value: "": must contain only alpha-numeric characters (a-z, 0-9), and hyphens (-)
* spec.containers[2].startupProbe.httpGet.port: Invalid value: "": must contain at least one letter (a-z)
```

Reply:
FWIW, I pushed a change last night requires the port for both handlers and probers.